File: web/src/frontends/classic/app/resources/health.ts
import { queryOptions } from '@tanstack/vue-query'
import type { MaybeRefOrGetter } from 'vue'

import type { ApiClient } from '@shared/http/client'
import type {
  HealthGroupDto,
  HealthProblemCredentialDto,
  HealthQuotaCredentialDto,
  HealthExpiringResetCreditDto,
  HealthRecoveryDto,
  HealthCredentialCountsDto,
  HealthAccessKeyCostLimitDto,
  RequestLogHealthDto,
  RuntimeHealthDto,
} from '@/api/control/types'
import { InvalidResponseError } from '@shared/http/errors'
import { controlQueryKeys } from '@/app/query-keys'
import { projectAccessKeyCostLimitRuleStatus } from './access-keys'

import {
  assertNoSecretLikeFields,
  projectArray,
  projectBoolean,
  projectEpochMilliseconds,
  projectEnum,
  projectNullableEpochMilliseconds,
  projectRecord,
  projectSafeInteger,
  projectString,
} from './projector'

export type {
  HealthGroupDto,
  HealthProblemCredentialDto,
  HealthQuotaCredentialDto,
  HealthExpiringResetCreditDto,
  HealthRecoveryDto,
  HealthCredentialCountsDto,
  RequestLogHealthDto,
  RuntimeHealthDto,
} from '@/api/control/types'

const countFields = ['credentials', 'available', 'cooldown', 'blacklisted'] as const
const healthFields = [
  'observed_at_ms',
  'version',
  'uptime_seconds',
  'snapshot_revision',
  'stats_window_seconds',
  'counts',
  'groups',
  'cooldown_credentials',
  'blacklisted_credentials',
  'low_quota_credentials',
  'expiring_reset_credits',
  'blocked_access_keys',
  'request_log',
] as const
const blockedAccessKeyFields = [
  'access_key_id',
  'name',
  'masked_key',
  'recoverable',
  'next_available_at_ms',
  'blocking_rules',
] as const
const quotaCredentialFields = [
  'credential_id',
  'group_id',
  'group_name',
  'remaining',
  'reset_at_ms',
] as const
const expiringResetCreditFields = [
  'credential_id',
  'group_id',
  'group_name',
  'count',
  'nearest_expires_at_ms',
] as const
const problemCredentialFields = [
  'credential_id',
  'group_id',
  'group_name',
  'cooldown_until_ms',
  'failure_count',
  'recent_success_count',
  'recent_problem_count',
  'consecutive_problem_count',
  'weight',
  'recovery',
  'identity',
  'last_failure_category',
  'last_status_code',
] as const
const requestLogFields = [
  'enqueued_total',
  'persisted_total',
  'dropped_not_running_total',
  'dropped_queue_full_total',
  'dropped_stopping_total',
  'dropped_persist_failed_total',
  'dropped_shutdown_total',
  'dropped_total',
  'write_failure_total',
  'access_quota_checkpoint_write_failure_total',
  'access_quota_checkpoint_degraded',
  'retention_delete_failure_total',
  'queue_depth',
  'queue_capacity',
  'last_write_failure_at_ms',
  'last_access_quota_checkpoint_write_failure_at_ms',
  'last_retention_failure_at_ms',
] as const
const requestLogCounterFields = [
  'enqueued_total',
  'persisted_total',
  'dropped_not_running_total',
  'dropped_queue_full_total',
  'dropped_stopping_total',
  'dropped_persist_failed_total',
  'dropped_shutdown_total',
  'dropped_total',
  'write_failure_total',
  'access_quota_checkpoint_write_failure_total',
  'retention_delete_failure_total',
  'queue_depth',
  'queue_capacity',
] as const
const recoveryModes = ['cooldown_expiry', 'validation_probe', 'configuration_required'] as const
const problemFailureCategories = [
  'rate_limited',
  'model_unavailable',
  'invalid_key',
  'upstream_host_error',
  'client_error',
  'downstream_cancel',
  'authentication_required',
  'ambiguous',
] as const
function invalidResponse(): never {
  throw new InvalidResponseError()
}

function projectNonBlankString(value: unknown): string {
  const result = projectString(value)
  if (result.trim().length === 0) invalidResponse()
  return result
}

export function projectHealthCounts(value: unknown): HealthCredentialCountsDto {
  const record = projectRecord(value)
  assertNoSecretLikeFields(record, countFields)
  const result = {
    credentials: projectSafeInteger(record.credentials, { minimum: 0 }),
    available: projectSafeInteger(record.available, { minimum: 0 }),
    cooldown: projectSafeInteger(record.cooldown, { minimum: 0 }),
    blacklisted: projectSafeInteger(record.blacklisted, { minimum: 0 }),
  }
  if (result.credentials !== result.available + result.cooldown + result.blacklisted) {
    invalidResponse()
  }
  return result
}

function projectHealthGroup(value: unknown): HealthGroupDto {
  const record = projectRecord(value)
  assertNoSecretLikeFields(record, ['id', 'name', 'enabled', 'counts'])
  const { model_cooldown, ...counts } = projectRecord(record.counts)
  return {
    id: projectSafeInteger(record.id, { minimum: 1 }),
    name: projectNonBlankString(record.name),
    enabled: projectBoolean(record.enabled),
    counts: {
      ...projectHealthCounts(counts),
      model_cooldown: projectSafeInteger(model_cooldown, { minimum: 0 }),
    },
  }
}

function projectRecovery(value: unknown): HealthRecoveryDto {
  const record = projectRecord(value)
  assertNoSecretLikeFields(record, ['automatic', 'mode', 'at_ms'])
  return {
    automatic: projectBoolean(record.automatic),
    mode: projectEnum(record.mode, recoveryModes),
    at_ms: projectNullableEpochMilliseconds(record.at_ms),
  }
}

function projectProblemCredential(value: unknown): HealthProblemCredentialDto {
  const record = projectRecord(value)
  assertNoSecretLikeFields(record, problemCredentialFields)
  const recovery = projectRecovery(record.recovery)
  const cooldownUntilMS = projectNullableEpochMilliseconds(record.cooldown_until_ms)

  if (recovery.mode === 'cooldown_expiry') {
    if (!recovery.automatic || recovery.at_ms !== cooldownUntilMS) invalidResponse()
  } else if (recovery.mode === 'validation_probe') {
    if (!recovery.automatic || cooldownUntilMS !== null || recovery.at_ms !== null) {
      invalidResponse()
    }
  } else if (recovery.automatic || cooldownUntilMS !== null || recovery.at_ms !== null) {
    invalidResponse()
  }

  return {
    credential_id: projectSafeInteger(record.credential_id, { minimum: 1 }),
    group_id: projectSafeInteger(record.group_id, { minimum: 1 }),
    group_name: projectNonBlankString(record.group_name),
    cooldown_until_ms: cooldownUntilMS,
    failure_count: projectSafeInteger(record.failure_count, { minimum: 0 }),
    recent_success_count: projectSafeInteger(record.recent_success_count, { minimum: 0 }),
    recent_problem_count: projectSafeInteger(record.recent_problem_count, { minimum: 0 }),
    consecutive_problem_count: projectSafeInteger(record.consecutive_problem_count, {
      minimum: 0,
    }),
    weight: projectSafeInteger(record.weight, { minimum: 0, maximum: 100 }),
    recovery,
    identity: projectString(record.identity),
    last_failure_category: projectEnum(record.last_failure_category, problemFailureCategories),
    last_status_code:
      record.last_status_code === null
        ? null
        : projectSafeInteger(record.last_status_code, { minimum: 100, maximum: 999 }),
  }
}

function projectQuotaCredential(value: unknown): HealthQuotaCredentialDto {
  const record = projectRecord(value)
  assertNoSecretLikeFields(record, quotaCredentialFields)
  const remaining = record.remaining
  if (
    typeof remaining !== 'number' ||
    !Number.isFinite(remaining) ||
    remaining < 0 ||
    remaining > 1
  ) {
    invalidResponse()
  }
  return {
    credential_id: projectSafeInteger(record.credential_id, { minimum: 1 }),
    group_id: projectSafeInteger(record.group_id, { minimum: 1 }),
    group_name: projectNonBlankString(record.group_name),
    remaining,
    reset_at_ms: projectEpochMilliseconds(record.reset_at_ms),
  }
}

function projectExpiringResetCredit(value: unknown): HealthExpiringResetCreditDto {
  const record = projectRecord(value)
  assertNoSecretLikeFields(record, expiringResetCreditFields)
  return {
    credential_id: projectSafeInteger(record.credential_id, { minimum: 1 }),
    group_id: projectSafeInteger(record.group_id, { minimum: 1 }),
    group_name: projectNonBlankString(record.group_name),
    count: projectSafeInteger(record.count, { minimum: 1 }),
    nearest_expires_at_ms: projectEpochMilliseconds(record.nearest_expires_at_ms),
  }
}

function projectBlockedAccessKey(value: unknown): HealthAccessKeyCostLimitDto {
  const record = projectRecord(value)
  assertNoSecretLikeFields(record, blockedAccessKeyFields)
  const recoverable = projectBoolean(record.recoverable)
  const nextAvailable = projectNullableEpochMilliseconds(record.next_available_at_ms)
  const blockingRules = projectArray(record.blocking_rules, projectAccessKeyCostLimitRuleStatus)
  if (
    blockingRules.length === 0 ||
    blockingRules.some((rule) => rule.status !== 'exhausted') ||
    (recoverable &&
      (nextAvailable === null || blockingRules.some((rule) => rule.kind !== 'periodic'))) ||
    (!recoverable &&
      (nextAvailable !== null || !blockingRules.some((rule) => rule.kind === 'total')))
  ) {
    invalidResponse()
  }
  return {
    access_key_id: projectSafeInteger(record.access_key_id, { minimum: 1 }),
    name: projectNonBlankString(record.name),
    masked_key: projectString(record.masked_key),
    recoverable,
    next_available_at_ms: nextAvailable,
    blocking_rules: blockingRules,
  }
}

function projectRequestLogHealth(value: unknown): RequestLogHealthDto {
  const record = projectRecord(value)
  assertNoSecretLikeFields(record, requestLogFields)
  const counters = Object.fromEntries(
    requestLogCounterFields.map((field) => [
      field,
      projectSafeInteger(record[field], { minimum: 0 }),
    ]),
  ) as Pick<
    RequestLogHealthDto,
    Exclude<
      keyof RequestLogHealthDto,
      | 'last_write_failure_at_ms'
      | 'last_access_quota_checkpoint_write_failure_at_ms'
      | 'last_retention_failure_at_ms'
      | 'access_quota_checkpoint_degraded'
    >
  >
  return {
    ...counters,
    access_quota_checkpoint_degraded: projectBoolean(record.access_quota_checkpoint_degraded),
    last_write_failure_at_ms: projectNullableEpochMilliseconds(record.last_write_failure_at_ms),
    last_access_quota_checkpoint_write_failure_at_ms: projectNullableEpochMilliseconds(
      record.last_access_quota_checkpoint_write_failure_at_ms,
    ),
    last_retention_failure_at_ms: projectNullableEpochMilliseconds(
      record.last_retention_failure_at_ms,
    ),
  }
}

export function projectRuntimeHealth(value: unknown): RuntimeHealthDto {
  const record = projectRecord(value)
  assertNoSecretLikeFields(record, healthFields)
  return {
    observed_at_ms: projectEpochMilliseconds(record.observed_at_ms),
    version: projectNonBlankString(record.version),
    uptime_seconds: projectSafeInteger(record.uptime_seconds, { minimum: 0 }),
    snapshot_revision: projectSafeInteger(record.snapshot_revision, { minimum: 1 }),
    stats_window_seconds: projectSafeInteger(record.stats_window_seconds, { minimum: 1 }),
    counts: projectHealthCounts(record.counts),
    groups: projectArray(record.groups, projectHealthGroup),
    cooldown_credentials: projectArray(record.cooldown_credentials, projectProblemCredential),
    blacklisted_credentials: projectArray(record.blacklisted_credentials, projectProblemCredential),
    low_quota_credentials: projectArray(record.low_quota_credentials, projectQuotaCredential),
    expiring_reset_credits: projectArray(record.expiring_reset_credits, projectExpiringResetCredit),
    blocked_access_keys: projectArray(record.blocked_access_keys, projectBlockedAccessKey),
    request_log: projectRequestLogHealth(record.request_log),
  }
}

export async function getRuntimeHealth(
  client: ApiClient,
  signal?: AbortSignal,
): Promise<RuntimeHealthDto> {
  return projectRuntimeHealth(await client.request('/api/health', { method: 'GET', signal }))
}

export function healthQueryOptions(
  client: ApiClient,
  intervalMs?: number,
  enabled?: MaybeRefOrGetter<boolean>,
) {
  return queryOptions({
    queryKey: controlQueryKeys.health(),
    queryFn: ({ signal }) => getRuntimeHealth(client, signal),
    refetchOnWindowFocus: false,
    ...(intervalMs !== undefined
      ? {
          refetchInterval: intervalMs,
          refetchIntervalInBackground: false,
        }
      : {}),
    // /api/health 不在 AccessKey 白名单里，调用方需要能按身份关掉这个查询。
    ...(enabled !== undefined ? { enabled } : {}),
  })
}

File: internal/control/health.go
package control

import (
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"gpt-load/internal/channel"
	"gpt-load/internal/health"
	app_errors "gpt-load/internal/platform/errors"
	"gpt-load/internal/platform/response"
	"gpt-load/internal/platform/version"
	"gpt-load/internal/requestlog"
	"gpt-load/internal/state"
)

type RequestLogStatsReader interface {
	Stats() requestlog.Stats
}

type healthCountsResponse struct {
	Credentials int `json:"credentials"`
	Available   int `json:"available"`
	Cooldown    int `json:"cooldown"`
	Blacklisted int `json:"blacklisted"`
}

type healthGroupCountsResponse struct {
	healthCountsResponse
	ModelCooldown int `json:"model_cooldown"`
}

type healthGroupResponse struct {
	ID      uint                      `json:"id"`
	Name    string                    `json:"name"`
	Enabled bool                      `json:"enabled"`
	Counts  healthGroupCountsResponse `json:"counts"`
}

type healthRecoveryResponse struct {
	Automatic bool   `json:"automatic"`
	Mode      string `json:"mode"`
	AtMS      *int64 `json:"at_ms"`
}

type healthProblemCredentialResponse struct {
	CredentialID            uint                   `json:"credential_id"`
	GroupID                 uint                   `json:"group_id"`
	GroupName               string                 `json:"group_name"`
	Identity                string                 `json:"identity"`
	LastFailureCategory     string                 `json:"last_failure_category"`
	LastStatusCode          *int                   `json:"last_status_code"`
	CooldownUntilMS         *int64                 `json:"cooldown_until_ms"`
	FailureCount            int                    `json:"failure_count"`
	RecentSuccessCount      uint64                 `json:"recent_success_count"`
	RecentProblemCount      uint64                 `json:"recent_problem_count"`
	ConsecutiveProblemCount uint64                 `json:"consecutive_problem_count"`
	Weight                  int                    `json:"weight"`
	Recovery                healthRecoveryResponse `json:"recovery"`
}

// healthQuotaCredentialResponse 描述额度即将耗尽的订阅凭据。
type healthQuotaCredentialResponse struct {
	CredentialID uint    `json:"credential_id"`
	GroupID      uint    `json:"group_id"`
	GroupName    string  `json:"group_name"`
	Identity     string  `json:"identity"`
	Remaining    float64 `json:"remaining"`
	ResetAtMS    int64   `json:"reset_at_ms"`
}

type healthExpiringResetCreditResponse struct {
	CredentialID       uint   `json:"credential_id"`
	GroupID            uint   `json:"group_id"`
	GroupName          string `json:"group_name"`
	Identity           string `json:"identity"`
	Count              int    `json:"count"`
	NearestExpiresAtMS int64  `json:"nearest_expires_at_ms"`
}

type requestLogHealthResponse struct {
	EnqueuedTotal                             uint64 `json:"enqueued_total"`
	PersistedTotal                            uint64 `json:"persisted_total"`
	DroppedNotRunningTotal                    uint64 `json:"dropped_not_running_total"`
	DroppedQueueFullTotal                     uint64 `json:"dropped_queue_full_total"`
	DroppedStoppingTotal                      uint64 `json:"dropped_stopping_total"`
	DroppedPersistFailedTotal                 uint64 `json:"dropped_persist_failed_total"`
	DroppedShutdownTotal                      uint64 `json:"dropped_shutdown_total"`
	DroppedTotal                              uint64 `json:"dropped_total"`
	WriteFailureTotal                         uint64 `json:"write_failure_total"`
	AccessQuotaCheckpointWriteFailureTotal    uint64 `json:"access_quota_checkpoint_write_failure_total"`
	AccessQuotaCheckpointDegraded             bool   `json:"access_quota_checkpoint_degraded"`
	RetentionDeleteFailureTotal               uint64 `json:"retention_delete_failure_total"`
	QueueDepth                                int    `json:"queue_depth"`
	QueueCapacity                             int    `json:"queue_capacity"`
	LastWriteFailureAtMS                      *int64 `json:"last_write_failure_at_ms"`
	LastAccessQuotaCheckpointWriteFailureAtMS *int64 `json:"last_access_quota_checkpoint_write_failure_at_ms"`
	LastRetentionFailureAtMS                  *int64 `json:"last_retention_failure_at_ms"`
}

type runtimeHealthResponse struct {
	ObservedAtMS           int64                               `json:"observed_at_ms"`
	Version                string                              `json:"version"`
	UptimeSeconds          int64                               `json:"uptime_seconds"`
	SnapshotRevision       uint64                              `json:"snapshot_revision"`
	StatsWindowSeconds     int64                               `json:"stats_window_seconds"`
	Counts                 healthCountsResponse                `json:"counts"`
	Groups                 []healthGroupResponse               `json:"groups"`
	CooldownCredentials    []healthProblemCredentialResponse   `json:"cooldown_credentials"`
	BlacklistedCredentials []healthProblemCredentialResponse   `json:"blacklisted_credentials"`
	LowQuotaCredentials    []healthQuotaCredentialResponse     `json:"low_quota_credentials"`
	ExpiringResetCredits   []healthExpiringResetCreditResponse `json:"expiring_reset_credits"`
	BlockedAccessKeys      []healthAccessKeyCostLimitResponse  `json:"blocked_access_keys"`
	RequestLog             requestLogHealthResponse            `json:"request_log"`
}

type healthAccessKeyCostLimitResponse struct {
	AccessKeyID       uint                           `json:"access_key_id"`
	Name              string                         `json:"name"`
	MaskedKey         string                         `json:"masked_key"`
	Recoverable       bool                           `json:"recoverable"`
	NextAvailableAtMS *int64                         `json:"next_available_at_ms"`
	BlockingRules     []AccessKeyCostLimitRuleStatus `json:"blocking_rules"`
}

const (
	// healthLowQuotaRemainingRatio 是「额度快用完」的唯一阈值来源。
	// 与管理 UI 账号卡的 danger 档保持一致，避免同一句结论在两处算出不同答案。
	healthLowQuotaRemainingRatio = 0.3
)

type healthBucket string

const (
	healthBucketAvailable   healthBucket = "available"
	healthBucketCooldown    healthBucket = "cooldown"
	healthBucketBlacklisted healthBucket = "blacklisted"
	healthBucketDisabled    healthBucket = "disabled"
)

func classifyHealthKey(
	group state.GroupCatalogView,
	key state.CredentialRuntimeView,
	now time.Time,
) healthBucket {
	if !group.Enabled ||
		(group.WeightManual != nil && *group.WeightManual == 0) ||
		key.Status != state.CredentialStatusActive ||
		(key.AuthState != "" && key.AuthState != state.CredentialAuthStateReady) ||
		(key.WeightManual != nil && *key.WeightManual == 0) {
		return healthBucketDisabled
	}
	switch key.RuntimeState(now) {
	case state.CredentialRuntimeBlacklisted:
		return healthBucketBlacklisted
	case state.CredentialRuntimeCooldown:
		return healthBucketCooldown
	default:
		return healthBucketAvailable
	}
}

func addHealthCount(counts *healthCountsResponse, bucket healthBucket) {
	switch bucket {
	case healthBucketAvailable:
		counts.Credentials++
		counts.Available++
	case healthBucketCooldown:
		counts.Credentials++
		counts.Cooldown++
	case healthBucketBlacklisted:
		counts.Credentials++
		counts.Blacklisted++
	}
}

func cloneInt(value *int) *int {
	if value == nil {
		return nil
	}
	cloned := *value
	return &cloned
}

func optionalHealthStatusCode(value int) *int {
	if value < 100 || value > 999 {
		return nil
	}
	cloned := value
	return &cloned
}

// healthProblemCredentialIdentity 返回健康问题列表里凭据的展示身份：
// API 密钥仍是掩码，订阅账号给完整邮箱，与凭据卡片、日志的既有约定一致；
// 拿不到邮箱时回退到 "Subscription #ID"。access/refresh token 绝不进入返回值。
func (service *Service) healthProblemCredentialIdentity(
	ciphertexts map[uint]string,
	credentialID uint,
	channelID channel.ID,
	connectionType string,
) (string, error) {
	if service == nil || service.encryption == nil || credentialID == 0 {
		return "", fmt.Errorf(
			"map runtime health problem key: %w",
			app_errors.ErrInternalServer,
		)
	}
	ciphertext, exists := ciphertexts[credentialID]
	if !exists || ciphertext == "" {
		return "", fmt.Errorf(
			"map runtime health problem key %d: ciphertext unavailable: %w",
			credentialID,
			app_errors.ErrInternalServer,
		)
	}
	plaintext, err := service.encryption.Decrypt(ciphertext)
	if err != nil {
		return "", fmt.Errorf(
			"map runtime health problem key %d: decrypt credential: %v: %w",
			credentialID,
			err,
			app_errors.ErrInternalServer,
		)
	}
	expectedConnection, known := service.channelRegistry.ConnectionType(channelID)
	if !known || expectedConnection != strings.TrimSpace(connectionType) {
		plaintext = ""
		return "", fmt.Errorf("map runtime health credential %d: channel connection mismatch: %w", credentialID, app_errors.ErrInternalServer)
	}
	if driver, bound := service.subscriptions.Driver(channelID); bound {
		credential, parseErr := driver.Parse([]byte(plaintext))
		plaintext = ""
		if parseErr != nil {
			return "", fmt.Errorf("map runtime health subscription credential %d: %w", credentialID, app_errors.ErrInternalServer)
		}
		if email := strings.TrimSpace(credential.Account().Email); email != "" {
			return email, nil
		}
		return fmt.Sprintf("Subscription #%d", credentialID), nil
	}
	validated, err := normalizeStoredCredential(service.channelRegistry, channelID, plaintext)
	if err != nil {
		return "", fmt.Errorf(
			"map runtime health problem credential %d: validate credential: %v: %w",
			credentialID,
			err,
			app_errors.ErrInternalServer,
		)
	}
	return maskCanonicalCredential(validated.CanonicalJSON())
}

func (service *Service) RuntimeHealth() (runtimeHealthResponse, error) {
	observation, err := service.captureRuntimeHealthObservation()
	if err != nil {
		return runtimeHealthResponse{}, err
	}
	if service.stats == nil || service.requestLogStats == nil {
		return runtimeHealthResponse{}, fmt.Errorf(
			"runtime health dependencies unavailable: %w",
			app_errors.ErrInternalServer,
		)
	}
	observedAtMS, err := safeEpochMilliseconds(observation.observedAt)
	if err != nil {
		return runtimeHealthResponse{}, fmt.Errorf("map runtime health observed_at_ms: %w", err)
	}
	result := runtimeHealthResponse{
		ObservedAtMS:           observedAtMS,
		SnapshotRevision:       observation.snapshot.Revision,
		StatsWindowSeconds:     int64(health.StatsWindow / time.Second),
		Groups:                 []healthGroupResponse{},
		CooldownCredentials:    []healthProblemCredentialResponse{},
		BlacklistedCredentials: []healthProblemCredentialResponse{},
		LowQuotaCredentials:    []healthQuotaCredentialResponse{},
		ExpiringResetCredits:   []healthExpiringResetCreditResponse{},
		BlockedAccessKeys:      []healthAccessKeyCostLimitResponse{},
	}
	groupIDs := make([]uint, 0, len(observation.snapshot.GroupCatalog))
	for groupID := range observation.snapshot.GroupCatalog {
		groupIDs = append(groupIDs, groupID)
	}
	sort.Slice(groupIDs, func(i, j int) bool { return groupIDs[i] < groupIDs[j] })
	groupIndexes := make(map[uint]int, len(groupIDs))
	for _, groupID := range groupIDs {
		group := observation.snapshot.GroupCatalog[groupID]
		groupIndexes[groupID] = len(result.Groups)
		result.Groups = append(result.Groups, healthGroupResponse{
			ID: group.ID, Name: group.Name, Enabled: group.Enabled,
		})
	}
	// 仅解密实际进入问题列表的凭据，同一账号的多个问题共用展示身份。
	identities := make(map[uint]string)
	identityFor := func(credentialID, groupID uint) (string, error) {
		if identity, exists := identities[credentialID]; exists {
			return identity, nil
		}
		group := observation.snapshot.Groups[groupID]
		identity, err := service.healthProblemCredentialIdentity(
			observation.credentialCiphertexts,
			credentialID,
			group.ChannelID,
			group.ConnectionType,
		)
		if err != nil {
			return "", err
		}
		identities[credentialID] = identity
		return identity, nil
	}
	resetCreditTargets := make(map[uint]healthResetCreditTarget)
	for _, key := range observation.keys {
		group := observation.snapshot.GroupCatalog[key.GroupID]
		index := groupIndexes[key.GroupID]
		bucket := classifyHealthKey(group, key, observation.observedAt)
		if bucket != healthBucketDisabled {
			resetCreditTargets[key.ID] = healthResetCreditTarget{
				groupID: key.GroupID, groupName: group.Name,
			}
		}
		if hasModelCooldown(key.ModelCooldowns, observation.observedAt) {
			result.Groups[index].Counts.ModelCooldown++
		}
		addHealthCount(&result.Counts, bucket)
		addHealthCount(&result.Groups[index].Counts.healthCountsResponse, bucket)
		// 额度只用于管理面展示，不参与健康分桶或调度；低额度凭据在这里单列提示。
		if bucket == healthBucketAvailable || bucket == healthBucketCooldown {
			if remaining := key.ObservedQuotaRemaining(); remaining != nil &&
				*remaining <= healthLowQuotaRemainingRatio {
				resetAtMS, err := safeEpochMilliseconds(key.QuotaResetAt)
				if err != nil {
					return runtimeHealthResponse{}, fmt.Errorf(
						"map low quota credential %d reset_at_ms: %w", key.ID, err,
					)
				}
				identity, err := identityFor(key.ID, key.GroupID)
				if err != nil {
					return runtimeHealthResponse{}, err
				}
				result.LowQuotaCredentials = append(
					result.LowQuotaCredentials,
					healthQuotaCredentialResponse{
						CredentialID: key.ID,
						GroupID:      key.GroupID,
						GroupName:    group.Name,
						Identity:     identity,
						Remaining:    *remaining,
						ResetAtMS:    resetAtMS,
					},
				)
			}
		}
		if bucket != healthBucketCooldown && bucket != healthBucketBlacklisted {
			continue
		}
		identity, err := identityFor(key.ID, key.GroupID)
		if err != nil {
			return runtimeHealthResponse{}, err
		}
		stats := service.stats.Snapshot(key.ID, observation.observedAt)
		lastFailureCategory := stats.LastFailureCategory
		lastStatusCode := stats.LastStatusCode
		if lastFailureCategory == health.FailureCategoryOK {
			lastFailureCategory = health.FailureCategoryAmbiguous
			lastStatusCode = 0
		}
		detail := healthProblemCredentialResponse{
			CredentialID:            key.ID,
			GroupID:                 key.GroupID,
			GroupName:               group.Name,
			Identity:                identity,
			LastFailureCategory:     lastFailureCategory.String(),
			LastStatusCode:          optionalHealthStatusCode(lastStatusCode),
			FailureCount:            key.FailureCount,
			RecentSuccessCount:      stats.Success,
			RecentProblemCount:      stats.Problem,
			ConsecutiveProblemCount: stats.ConsecutiveProblem,
			Weight:                  state.ConfiguredWeight(key.WeightManual),
		}
		if bucket == healthBucketCooldown {
			cooldownUntilMS, err := optionalSafeEpochMilliseconds(key.CooldownUntil)
			if err != nil {
				return runtimeHealthResponse{}, fmt.Errorf(
					"map runtime health cooldown_until_ms: %w",
					err,
				)
			}
			detail.CooldownUntilMS = cooldownUntilMS
			detail.Recovery = healthRecoveryResponse{
				Automatic: true,
				Mode:      "cooldown_expiry",
				AtMS:      cooldownUntilMS,
			}
			result.CooldownCredentials = append(result.CooldownCredentials, detail)
		} else {
			detail.Recovery = healthRecoveryResponse{Mode: "configuration_required"}
			if service.executor != nil && service.channelRegistry != nil {
				if validationGroup, exists := observation.snapshot.Groups[key.GroupID]; exists {
					if _, valid := buildGroupValidationTarget(validationGroup); valid {
						detail.Recovery = healthRecoveryResponse{Automatic: true, Mode: "validation_probe"}
					}
				}
			}
			result.BlacklistedCredentials = append(result.BlacklistedCredentials, detail)
		}
	}
	result.ExpiringResetCredits, err = service.loadExpiringResetCredits(
		resetCreditTargets,
		observation.observedAt,
	)
	if err != nil {
		return runtimeHealthResponse{}, err
	}
	for index := range result.ExpiringResetCredits {
		credit := &result.ExpiringResetCredits[index]
		identity, err := identityFor(credit.CredentialID, credit.GroupID)
		if err != nil {
			return runtimeHealthResponse{}, err
		}
		credit.Identity = identity
	}
	if observation.accessQuotaViews != nil {
		accessKeyIDs := make([]uint, 0, len(observation.snapshot.AccessKeysByID))
		for accessKeyID := range observation.snapshot.AccessKeysByID {
			accessKeyIDs = append(accessKeyIDs, accessKeyID)
		}
		sort.Slice(accessKeyIDs, func(i, j int) bool { return accessKeyIDs[i] < accessKeyIDs[j] })
		for _, accessKeyID := range accessKeyIDs {
			accessKey := observation.snapshot.AccessKeysByID[accessKeyID]
			if accessKey.Status != state.AccessKeyStatusActive {
				continue
			}
			view := observation.accessQuotaViews[accessKeyID]
			if view.Allowed {
				continue
			}
			if !validAccessKeyPrefix(accessKey.KeyPrefix) || !validAccessKeySuffix(accessKey.KeySuffix) {
				return runtimeHealthResponse{}, fmt.Errorf(
					"map blocked access key %d suffix: %w",
					accessKeyID,
					app_errors.ErrInternalServer,
				)
			}
			status := mapAccessKeyCostLimitStatus(view)
			result.BlockedAccessKeys = append(result.BlockedAccessKeys, healthAccessKeyCostLimitResponse{
				AccessKeyID: accessKeyID, Name: accessKey.Name,
				MaskedKey:         maskedAccessKey(accessKey.KeyPrefix, accessKey.KeySuffix),
				Recoverable:       status.Recoverable,
				NextAvailableAtMS: cloneCostLimitMilliseconds(status.NextAvailableAtMS),
				BlockingRules:     blockingCostLimitRuleStatuses(status),
			})
		}
	}
	requestLog, err := mapRequestLogHealth(service.requestLogStats.Stats())
	if err != nil {
		return runtimeHealthResponse{}, fmt.Errorf("map request log health: %w", err)
	}
	result.RequestLog = requestLog
	return result, nil
}

func mapRequestLogHealth(stats requestlog.Stats) (requestLogHealthResponse, error) {
	for _, value := range []uint64{
		stats.EnqueuedTotal,
		stats.PersistedTotal,
		stats.DroppedNotRunningTotal,
		stats.DroppedQueueFullTotal,
		stats.DroppedStoppingTotal,
		stats.DroppedPersistFailedTotal,
		stats.DroppedShutdownTotal,
		stats.DroppedTotal,
		stats.WriteFailureTotal,
		stats.AccessQuotaCheckpointWriteFailureTotal,
		stats.RetentionDeleteFailureTotal,
	} {
		if value > uint64(maxSafeInteger) {
			return requestLogHealthResponse{}, fmt.Errorf("map request log health: unsafe counter")
		}
	}
	if stats.QueueDepth < 0 || uint64(stats.QueueDepth) > uint64(maxSafeInteger) ||
		stats.QueueCapacity < 0 || uint64(stats.QueueCapacity) > uint64(maxSafeInteger) {
		return requestLogHealthResponse{}, fmt.Errorf("map request log health: unsafe queue")
	}
	lastWriteFailureAtMS, err := optionalSafeEpochMilliseconds(stats.LastWriteFailureAt)
	if err != nil {
		return requestLogHealthResponse{}, fmt.Errorf("map last write failure timestamp: %w", err)
	}
	lastAccessQuotaCheckpointWriteFailureAtMS, err := optionalSafeEpochMilliseconds(
		stats.LastAccessQuotaCheckpointWriteFailureAt,
	)
	if err != nil {
		return requestLogHealthResponse{}, fmt.Errorf(
			"map last access quota checkpoint write failure timestamp: %w",
			err,
		)
	}
	lastRetentionFailureAtMS, err := optionalSafeEpochMilliseconds(stats.LastRetentionFailureAt)
	if err != nil {
		return requestLogHealthResponse{}, fmt.Errorf("map last retention failure timestamp: %w", err)
	}
	return requestLogHealthResponse{
		EnqueuedTotal:                             stats.EnqueuedTotal,
		PersistedTotal:                            stats.PersistedTotal,
		DroppedNotRunningTotal:                    stats.DroppedNotRunningTotal,
		DroppedQueueFullTotal:                     stats.DroppedQueueFullTotal,
		DroppedStoppingTotal:                      stats.DroppedStoppingTotal,
		DroppedPersistFailedTotal:                 stats.DroppedPersistFailedTotal,
		DroppedShutdownTotal:                      stats.DroppedShutdownTotal,
		DroppedTotal:                              stats.DroppedTotal,
		WriteFailureTotal:                         stats.WriteFailureTotal,
		AccessQuotaCheckpointWriteFailureTotal:    stats.AccessQuotaCheckpointWriteFailureTotal,
		AccessQuotaCheckpointDegraded:             stats.AccessQuotaCheckpointDegraded,
		RetentionDeleteFailureTotal:               stats.RetentionDeleteFailureTotal,
		QueueDepth:                                stats.QueueDepth,
		QueueCapacity:                             stats.QueueCapacity,
		LastWriteFailureAtMS:                      lastWriteFailureAtMS,
		LastAccessQuotaCheckpointWriteFailureAtMS: lastAccessQuotaCheckpointWriteFailureAtMS,
		LastRetentionFailureAtMS:                  lastRetentionFailureAtMS,
	}, nil
}

func (server *Server) handleRuntimeHealth(c *gin.Context) {
	result, err := server.service.RuntimeHealth()
	if err != nil {
		writeServiceError(c, "runtime_health", err)
		return
	}
	now := server.now().UTC()
	uptime := now.Sub(server.startedAt)
	if uptime < 0 {
		uptime = 0
	}
	result.Version = version.Version
	result.UptimeSeconds = int64(uptime / time.Second)
	response.SuccessI18n(c, "common.success", result)
}
